src/agent/src/device.rs

```diff
@@ -762,6 +836,45 @@ async fn vfio_device_handler(device: &Device, sandbox: &Arc<Mutex<Sandbox>>) ->
     })
 }
 
+// The VFIO AP (Adjunct Processor) device handler takes all the APQNs provided as device options
+// and awaits them. It sets the minimum AP rescan time of 5 seconds and temporarily adds that
+// amoutn to the hotplug timeout.
+#[cfg(target_arch = "s390x")]
+#[instrument]
+async fn vfio_ap_device_handler(
+    device: &Device,
+    sandbox: &Arc<Mutex<Sandbox>>,
+) -> Result<SpecUpdate> {
+    for ap_opt in device.options.iter() {
+        let opt = ap_opt.strip_prefix("ap:").ok_or_else(|| {
+            anyhow!(
+                "VFIO AP device option {} unexpectedly missed `ap:` indication",
+                ap_opt
+            )
+        })?;
+
+        let mut config = AGENT_CONFIG.write().await;
+        let hotplug_timeout = config.hotplug_timeout;
+        warn!(sl!(), "Temporarily adding {} seconds to hotplug timeout for AP devices due to rescan intervals", AP_CONFIG_TIME_SECONDS);
+        config.hotplug_timeout += time::Duration::from_secs(AP_CONFIG_TIME_SECONDS as u64 + 5);
+        // Drop write lock on config so that the uevent matcher can read from it
+        drop(config);
+        fs::write(AP_CONFIG_TIME_PATH, AP_CONFIG_TIME_SECONDS.to_string())?;
+
+        for apqn in opt.split(',') {
+            wait_for_ap_device(sandbox, ap::Device::from_str(apqn)?).await?;
+        }
+
+        AGENT_CONFIG.write().await.hotplug_timeout = hotplug_timeout;
```

Comment:
Should this be `config.hotplug_timout` after the increment but before the `drop()`?

Reply:
No, this is just the `hotplug_timeout` as it was before we updated it. This is a regular backup-and-restore.